patroni/quorum.py

```diff
@@ -0,0 +1,187 @@
+    transitioning to sync replication, i.e. `quorum, voters = 1, {'leader'}`, then the special leader semantics
+    mean that we could set `numsync, sync = 2, {'leader', 's1', 's2'}` in one step without worrying of the case that
+    s1, s2 have latest xlog, but not leader. The benefit seems too small to warrant adding any complexity.
+    """
+    def __init__(self, quorum, voters, numsync, sync, active, sync_wanted):
+        self.quorum = quorum
+        self.voters = set(voters)
+        self.numsync = numsync
+        self.sync = set(sync)
+        self.active = active
+        self.sync_wanted = sync_wanted
+
+    def check_invariants(self):
+        if self.quorum and not (len(self.voters | self.sync) < self.quorum + self.numsync):
+            raise QuorumError("Quorum and sync not guaranteed to overlap: nodes %d >= quorum %d + sync %d" %
+                              (len(self.voters | self.sync), self.quorum, self.numsync))
+        if not (self.voters <= self.sync or self.sync <= self.voters):
+            raise QuorumError("Mismatched sets: quorum only=%s sync only=%s" %
+                              (self.voters - self.sync, self.sync - self.voters))
+
+    def quorum_update(self, quorum, voters):
+        if quorum < 1:
+            raise QuorumError("Quorum %d < 0 of (%s)" % (quorum, voters))
+        self.quorum = quorum
+        self.voters = voters
+        self.check_invariants()
+        logger.debug('quorum %s %s', self.quorum, self.voters)
+        return 'quorum', self.quorum, self.voters
+
+    def sync_update(self, numsync, sync):
+        self.numsync = numsync
+        self.sync = sync
+        self.check_invariants()
+        logger.debug('sync %s %s', self.numsync, self.sync)
+        return 'sync', self.numsync, self.sync
+
+    def __iter__(self):
+        transitions = list(self._generate_transitions())
+        # Merge 2 transitions of the same type to a single one. This is always safe because skipping the first
+        # transition is equivalent to no one observing the intermediate state.
+        for cur_transition, next_transition in zip(transitions, transitions[1:]+[None]):
+            if next_transition and cur_transition[0] == next_transition[0]:
+                continue
+            yield cur_transition
+
+    def _generate_transitions(self):
+        logger.debug("Quorum state: quorum %s, voters %s, numsync %s, sync %s, active %s, sync_wanted %s",
+                     self.quorum, self.voters, self.numsync, self.sync, self.active, self.sync_wanted)
+        self.check_invariants()
+
+        # Handle non steady state cases
+        if self.sync < self.voters:
+            logger.debug("Case 1: synchronous_standby_names subset of DCS state")
+            # Case 1: quorum is superset of sync nodes. In the middle of changing quorum.
+            # Evict from quorum dead nodes that are not being synced.
+            remove_from_quorum = self.voters - (self.sync | self.active)
```

Comment:
Can it happen at this point that `self.sync` still contains some node that ceased to be active and  therefore must be evicted from `voters` ? In other words, why do we subtract here the union and not `self.sync & self.active` or simply `self.active` ?